### PR TITLE
fix: correctly use mixed precision with multi-GPU in PyTorchTrial [DET-3285]

### PR DESCRIPTION
**Pull request by aaron276h**
## Description
When performing mixed precision multi-GPU training, we need to finalize gradient update communication prior to un-scaling.

## Test Plan
Re-enabled test and tested manually,

### CODE REVIEW

File: harness/determined/pytorch/_pytorch_trial.py [hunk 1]
```diff
@@ -389,11 +389,15 @@ def _train_for_step(self, step_id: int, batches_per_step: int) -> workload.Respo
             if self.use_amp():
                 with apex.amp.scale_loss(loss, self.context.optimizer) as scaled_loss:
                     scaled_loss.backward()
+                    if self.hvd_config.use and communicate_and_update:
+                        # When using horovod, we need to finish communicating gradient
+                        # updates before they are unscaled which happens when.
```
Comment (brain-good) [inline harness/determined/pytorch/_pytorch_trial.py:394]: blocking: need to clarify the end of this comment a bit. Didn't quite finish saying when it actually happens.

File: harness/determined/pytorch/_pytorch_trial.py [hunk 2]
```diff
@@ -389,11 +389,15 @@ def _train_for_step(self, step_id: int, batches_per_step: int) -> workload.Respo
...
+                        self.context.optimizer.synchronize()
             else:
                 loss.backward()
 
             if communicate_and_update:
-                if self.hvd_config.use:
+                if self.hvd_config.use and not self.use_amp():
                     self.context.optimizer.synchronize()
```
Comment (brain-good) [inline harness/determined/pytorch/_pytorch_trial.py:401]: non-blocking: I slightly prefer having this line as part of the `else` above. Since it takes a bit of a careful reading to realize that we are not calling synchronize twice.

File: harness/determined/pytorch/_pytorch_trial.py (hunk 1 above)
Comment (aaron276h) [reply harness/determined/pytorch/_pytorch_trial.py]: good catch, i just left off mid-sentence

File: harness/determined/pytorch/_pytorch_trial.py (hunk 2 above)
Comment (aaron276h) [reply harness/determined/pytorch/_pytorch_trial.py:403]: Moved it around, still not a great layout but I do agree its bit easier to follow now.
